### PR TITLE
[Kernel] Add onednn W8A8 compute kernels.

### PR DESCRIPTION
**Pull request by xiangzez**
This PR adds a W8A8 path which uses AMX_INT8 to compute for w8a8 weight type.

This path is faster than bf16/int8 for first token, slightly slower than int8 for next token(bs=1), faster than int8 for next token(bs>1).

### CODE REVIEW

File: src/utils/matmul_helper.h [hunk 1]
```diff
@@ -1541,11 +1626,230 @@ class MMHelper {
...
+        matmul_args.insert({DNNL_ARG_DST, output_mem});
+        t1.release();
+
+        // Executions.
+        TimeLine t2("onednn_gemm_s8s8s32.execute_primitive");
+        matmul_prim->execute(get_dnnl_stream(), matmul_args);
+        get_dnnl_stream().wait();
+    }
+
+    static void onednn_amxint8_f32s8f32_compute(bool transA, int M, int N, int K, float alpha, const float *A, int lda,
```
Comment (changqi1) [inline src/utils/matmul_helper.h:1683]: rename: onednn_amx_gemm_f32s8f32_compute

File: src/utils/matmul_helper.h [hunk 2]
```diff
@@ -1541,11 +1626,230 @@ class MMHelper {
         get_dnnl_stream().wait();
     }
 
-private:
+    static void onednn_gemm_s8s8s32(bool transA, int M, int N, int K, float alpha, const int8_t *A, int lda,
```
Comment (changqi1) [inline src/utils/matmul_helper.h:1629]: rename: onednn_amx_gemm_s8s8s32_compute

File: CMakeLists.txt [hunk 3]
```diff
@@ -112,6 +112,7 @@ add_definitions(-DAVX512_FP16_WEIGHT_ONLY_FP16=true)
 add_definitions(-DAVX512_BF16_WEIGHT_ONLY_BF16=true)
 # add_definitions(-DAVX512_FP32_WEIGHT_ONLY_INT8=true)
 add_definitions(-DAVX512_FP16_WEIGHT_ONLY_INT8=true)
+# add_definitions(-DAMX_INT8=true)
```
Comment (changqi1) [inline CMakeLists.txt:115]: // # add_definitions(-DAMX_INT8_W8A8=true)

File: src/utils/matmul_helper.h [hunk 4]
```diff
@@ -1541,11 +1626,230 @@ class MMHelper {
...
 
     static void set_amx_data_type(dnnl::memory::format_tag tag) { get_amx_data_type() = tag; }
+
+    static void quantize_s8(
+            int M, int N, const float *src, int lda, int8_t *dst, int ldb, float *scale, float *zero, float *sum) {
+#pragma omp parallel for
+        for (int i = 0; i < M; i++) {
+            __m512 vmax = _mm512_loadu_ps(src + i * lda);
+            __m512 vmin = vmax;
+            for (int j = 16; j < N; j += 16) {
```
[the remark this answers is not in the document]
Comment (xiangzez) [reply src/utils/matmul_helper.h:1730]: See line 1686

File: src/utils/matmul_helper.h [hunk 5]
```diff
@@ -1541,11 +1626,230 @@ class MMHelper {
...
+            return _mm512_mul_ps(v, vres);
+        };
+        auto silu = [](__m512 &v, int row, int col) {
+            __m512 vone = _mm512_set1_ps(1.0f);
+            __m512 vp = BertUtil::vexp(v);
+            __m512 vrecip = _mm512_rcp14_ps(vp + vone);
+            return vp * vrecip * v;
+        };
+
+        switch (kind) {
```
[the remark this answers is not in the document]
Comment (xiangzez) [reply src/utils/matmul_helper.h:1845]: I think the perf impact of one switch case is negligible?

**Comment by changqi1** [comment]
Please refer BF16 logic to write INT8 logic. But need to use the same pack function for onednn and xdnn.
And the Macro AMX_INT8_W8A8 may be unused.

```bash.
        // BF16
        else if constexpr (std::is_same_v<WeiT, bfloat16_t>) {
#ifdef AVX512_FP32_WEIGHT_ONLY_BF16
            TimeLine t("xdnn_sgemm_f32bf16f32_compute");
            xdnn_sgemm_f32bf16f32_compute(transA, M, N, K, alpha, A, lda, (const XDNN_UINT4x2 *)packedB, beta, C, ldc);
#elif defined(AVX512_BF16_WEIGHT_ONLY_BF16)
            if (M > USE_AMX_M) {
                TimeLine t("onednn_amx_sgemm_f32bf16f32_compute");
                onednn_amx_sgemm_f32bf16f32_compute(transA, M, N, K, alpha, A, lda, packedB, beta, C, ldc);
            } else {
                TimeLine t("xdnn_bgemm_f32bf16f32_compute");
                xdnn_bgemm_f32bf16f32_compute(transA, M, N, K, alpha, A, lda, (const XDNN_BF16 *)packedB, beta, C, ldc);
            }
#else
            printf("%s:%d: Need to define WEIGHT_ONLY_BF16 kernel data type.\n", __FILE__, __LINE__);
            exit(-1);
#endif
        }
        // INT8
        else if constexpr (std::is_same_v<WeiT, int8_t>) {
#ifdef AVX512_FP32_WEIGHT_ONLY_INT8
            TimeLine t("xdnn_sgemm_f32s8f32_compute");
            xdnn_sgemm_f32s8f32_compute(transA, M, N, K, alpha, A, lda, packedB, scaleB, zeroB, beta, C, ldc);
#elif defined(AVX512_FP16_WEIGHT_ONLY_INT8)
            TimeLine t("xdnn_hgemm_f32s8f32_compute");
            xdnn_hgemm_f32s8f32_compute(transA, M, N, K, alpha, A, lda, packedB, scaleB, zeroB, beta, C, ldc);
#elif defined(AMX_INT8_W8A8)
            TimeLine t("onednn_amx_gemm_f32s8f32_compute");
            onednn_amx_gemm_f32s8f32_compute(transA, M, N, K, alpha, A, lda, packedB, scaleB, zeroB, sumB, beta, C, ldc,
                    nullptr, nullptr, 0, 0.0f, matmul_kinds::Basic);
#else
            printf("%s:%d: Need to define WEIGHT_ONLY_INT8 kernel data type.\n", __FILE__, __LINE__);
            exit(-1);
```

File: src/models/models.cpp [hunk 6]
```diff
@@ -194,6 +194,7 @@ AutoModel::AutoModel(std::string modelPath, xft::DataType datatype) : Model() {
             case xft::DataType::fp16: setDecoder(new OptDecoder<float16_t>(modelPath)); break;
             case xft::DataType::bf16: setDecoder(new OptDecoder<bfloat16_t>(modelPath)); break;
             case xft::DataType::int8: setDecoder(new OptDecoder<int8_t>(modelPath)); break;
+            case xft::DataType::w8a8: setDecoder(new OptDecoder<w8a8_t>(modelPath)); break;
```
Comment (changqi1) [inline src/models/models.cpp:197]: need to add more config after case xft::DataType::bf16_int8:

case xft::DataType::bf16_w8a8:
    setDecoder(new HybridModel<OptDecoder, bfloat16_t, w8a8_t>(modelPath));
    break;

need to add more config after case xft::DataType::bf16_nf4:

case xft::DataType::w8a8_int8:
                setDecoder(new HybridModel<OptDecoder, w8a8_t, int8_t>(modelPath));
                break;
case xft::DataType::w8a8_int4:
                setDecoder(new HybridModel<OptDecoder, w8a8_t, uint4x2_t>(modelPath));
                break;
case xft::DataType::w8a8_nf4:
                setDecoder(new HybridModel<OptDecoder, w8a8_t, nf4x2_t>(modelPath));
                break;

**Comment by xiangzez** [comment]
PR is updated. @changqi1

**Review by changqi1 — APPROVED** [review]

**Pull request merged**